akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2ServerDemux.scala

```diff
@@ -209,14 +217,15 @@ private[http2] class Http2ServerDemux(http2Settings: Http2CommonSettings, initia
       setHandler(substreamIn, new InHandler {
         def onPush(): Unit = {
           val sub = grab(substreamIn)
-          pull(substreamIn)
           handleOutgoingCreated(sub)
+          // Once the incoming stream is handled, we decide if we need to pull more.
+          pullOutgoingSubStreams
```

Comment:
It is important to always invoke `pullOutgoingSubStreams` _after_ the `streamStates` collection has been manipulated (new streams added or removed), otherwise the `hasCapacity` condition changes meaning.